### PR TITLE
Add create_atoms function to LAMMPS.jl

### PR DESCRIPTION
**Pull request by jmeziere**
It is useful (at least for me) to be able to load atoms directly from Julia to LAMMPS. This pull request provides a wrapper around ```lammps_create_atoms``` to easily do this, similar to the [create_atoms](https://docs.lammps.org/Python_module.html#lammps.lammps.create_atoms) function in the python interface to LAMMPS.

### CODE REVIEW

File: src/LAMMPS.jl
Comment (Joroks) [inline src/LAMMPS.jl:275]: A more detailed explaination of the parameters would be nice. Specifically on what `dexpand` does, the format of the image flags, and an explaination that types specifies the individual atom types.

**Review by Joroks — COMMENTED** [review]
Thanks for the PR, this looks great!

One thing I'm not quite sure about, is the handling of datatypes not supported directly by lammps. Up until now we've just restricted the wrapper functions to be called with the correct datatype.

I understand that it would be convenient for the user if we'd just handle the type conversion where needed but I also don't think that it would be too much to ask to have the user do this themselves.

If we ultimately decide on handling the type conversion on our end, we'd have to make this consistent across our other methods as well. I also think that we shouldn't issue warnings in case of a type mismatch but rather just document this behavior in our docstrings.

@vchuravy what's your opinion on this?

**Comment by vchuravy** [comment]
Yeah I would rather check the DType and then error

File: src/LAMMPS.jl
Comment (jmeziere) [reply src/LAMMPS.jl]: Yep, fixed this.

File: src/LAMMPS.jl
Comment (jmeziere) [reply src/LAMMPS.jl]: I added some more info in the docstring. Let me know what you think and I can add or change anything you'd like.

File: src/LAMMPS.jl
Comment (Joroks) [reply src/LAMMPS.jl]: Is there a particular reason why you decided to determine the type of the property dynamically instead of just following the types specified in the [documentation](https://docs.lammps.org/Classes_atom.html#_CPPv4N9LAMMPS_NS4Atom7extractEPKc)? So `x` and `v` are always `Matrix{Float64}`, `type` is always `Vector{Int32}`, and `image` is always `Vector{IMAGEINT}` I think this would be a lot cleaner because this would allow us to enforce the correct type using a typeassert in the parameterlist.

To get the correct type of imageint etc. we could also just introduce

```julia
const BIGINT = API.lammps_extract_setting(C_NULL, "bigint") == 4 ? Int32 : Int64
const TAGINT = API.lammps_extract_setting(C_NULL, "tagint") == 4 ? Int32 : Int64
const IMAGEINT = API.lammps_extract_setting(C_NULL, "imageint") == 4 ? Int32 : Int64
```
as constants. (and probably export these as well)

Comment (jmeziere) [reply src/LAMMPS.jl]: I just was not sure that LAMMPS always had the same precision for floats. I thought maybe you could compile different versions with different precision. I will go ahead and change that.

What are BIGINT and TAGINT for, though? Just for completeness?

Comment (Joroks) [reply src/LAMMPS.jl]: Yes, they are just for completeness. They'll probably come in handy in other places though, so its nice to have them in case we need them later.

**Review by Joroks — APPROVED** [review]
@vchuravy could you take a quick look at this before I merge it?

File: src/LAMMPS.jl
Comment (vchuravy) [inline src/LAMMPS.jl:262]: This `v` being a raw pointer makes me nervous. It incentivies the user to write `pointer(v)` which is wrong in most circumstances. Same for image. 

To denote that this can be null, use `Nothing`

File: test/runtests.jl
Comment (vchuravy) [inline test/runtests.jl:333]:
```suggestion
        @test_throws ArgumentError create_atoms(lmp, x[1:2,:], id, types; v, image, bexpand=true) 
```

File: test/runtests.jl
Comment (vchuravy) [inline test/runtests.jl:336]: @jmeziere the edit suggestion applies for the subsequent lines as well.

File: src/LAMMPS.jl
Comment (Joroks) [inline src/LAMMPS.jl:284]:
```suggestion
    v = v == nothing ? C_NULL : v
    image = image == nothing ? C_NULL : image
```
`Base.unsafe_convert` can handle the type conversion

File: src/LAMMPS.jl
Comment (Joroks) [inline src/LAMMPS.jl:286]:
```suggestion
    API.lammps_create_atoms(lmp, numAtoms, id, types, x, v, image, bexpand ? 1 : 0)
```
We automatically convert `lmp` to `lmp.handle` using `Base.cconvert` and `Base.unsafe_convert`. This also checks that our handle isn't NULL

File: src/LAMMPS.jl
Comment (Joroks) [inline src/LAMMPS.jl:250]: The signature is out of date

**Review by Joroks — CHANGES_REQUESTED** [review]

File: src/LAMMPS.jl
Comment (jmeziere) [reply src/LAMMPS.jl]: The last commit should fix this

File: test/runtests.jl
Comment (jmeziere) [reply test/runtests.jl]: Sorry about that, just not paying enough attention.

File: test/runtests.jl
Comment (Joroks) [inline test/runtests.jl:311]:
```suggestion
        create_atoms(lmp, x, id, types; v, image, bexpand=true)
```

File: test/runtests.jl
Comment (Joroks) [inline test/runtests.jl:336]:
```suggestion
        @test_throws ArgumentError create_atoms(lmp, x, id, types; v=v[1:2,:], image, bexpand=true)
```

**Review by Joroks — COMMENTED** [review]
Please be a bit more thorough when revising your PR.
Testing that your code is working by typing `] test` in your REPL is allways a good idea too.

File: src/LAMMPS.jl
Comment (vchuravy) [inline src/LAMMPS.jl:84]: When/How can these change? I.e. could I load a different build of lammps where these are now different?

Or are the simply if my platform is 32bit or 64bit? If so we could just use the `Int` alias

Comment (vchuravy) [reply src/LAMMPS.jl:84]: So it can change when the LAMMPS library changes. I don't recall if `liblammps_path` is a compile-time constant and redefining invalidates the LAMMPS.jl cache.

Comment (vchuravy) [reply src/LAMMPS.jl:84]: We could in the `__init__` function check that the assumption hasn't changed. That would suffice for now.

Comment (Joroks) [reply src/LAMMPS.jl:84]: Changing the Library path requires a julia restart. At least according to our documentation.

Comment (Joroks) [reply src/LAMMPS.jl:84]: I'm not really sure how constant a `const` really ist though. Is it redefined on every julia restart or is it possible for it to persist?

Comment (vchuravy) [reply src/LAMMPS.jl:84]: It will be persisted into the cachefile that Julia emits during precompilation.

File: test/runtests.jl
Comment (jmeziere) [reply test/runtests.jl:311]: Sorry about that. I'll be sure to pay more attention.

File: src/LAMMPS.jl
Comment (jmeziere) [reply src/LAMMPS.jl:84]: Does precompilation in Julia include the linking step?

Since liblammp_path is the path of a shared object file, this would have to be linked.
Anecdotally, whenever I've set the library, it's always (as far as I can remember) run precompilation when I've reopened Julia.

Comment (Joroks) [reply src/LAMMPS.jl:84]: > We could in the __init__ function check that the assumption hasn't changed. That would suffice for now.

Would we then just error in the `__init__()` function if the integer sizes have changed?
Would this then cause cache invalidation i.e. would a restart of julia then trigger recompilation?
Or would the user have to manually invalidate the LAMMPS.jl cache somehow?

```julia
function __init__()
    BIGINT != (API.lammps_extract_setting(C_NULL, "bigint") == 4 ? Int32 : Int64) && 
         error("The size of the LAMMPS integer type BIGINT has changed! To fix this, restart julia.")
    TAGINT != (API.lammps_extract_setting(C_NULL, "tagint") == 4 ? Int32 : Int64) &&
        error("The size of the LAMMPS integer type TAGINT has changed! To fix this, restart julia.")
    IMAGEINT != (API.lammps_extract_setting(C_NULL, "tagint") == 4 ? Int32 : Int64) &&
        error("The size of the LAMMPS integer type IMAGEINT has changed! To fix this, restart julia.")
end
````

Comment (vchuravy) [reply src/LAMMPS.jl:84]: > Or would the user have to manually invalidate the LAMMPS.jl cache somehow?

Yeah the user would need to manually invalidate the cache, but at least we wouldn't get mysterious crashes

Comment (vchuravy) [reply src/LAMMPS.jl:84]: > Does precompilation in Julia include the linking step?

Julia uses `dlopen` and not static linking

Comment (Joroks) [reply src/LAMMPS.jl:84]: @jmeziere 

```suggestion
const BIGINT = API.lammps_extract_setting(C_NULL, "bigint") == 4 ? Int32 : Int64
const TAGINT = API.lammps_extract_setting(C_NULL, "tagint") == 4 ? Int32 : Int64
const IMAGEINT = API.lammps_extract_setting(C_NULL, "imageint") == 4 ? Int32 : Int64

function __init__()
    BIGINT != (API.lammps_extract_setting(C_NULL, "bigint") == 4 ? Int32 : Int64) &&
        error("The size of the LAMMPS integer type BIGINT has changed! To fix this, you need to manually invalidate the LAMMPS.jl cache.")
    TAGINT != (API.lammps_extract_setting(C_NULL, "tagint") == 4 ? Int32 : Int64) &&
         error("The size of the LAMMPS integer type TAGINT has changed! To fix this, you need to manually invalidate the LAMMPS.jl cache.")
     IMAGEINT != (API.lammps_extract_setting(C_NULL, "tagint") == 4 ? Int32 : Int64) &&
         error("The size of the LAMMPS integer type IMAGEINT has changed! To fix this, you need to manually invalidate the LAMMPS.jl cache.")
end
```

**Review by Joroks — COMMENTED** [review]
I think this should now be good to go.

File: src/LAMMPS.jl
Comment (jmeziere) [reply src/LAMMPS.jl:84]: I just pushed it up, so it should be good to go. Thanks for your patience with me. This is my first time going through the process.

File: src/LAMMPS.jl
Comment (vchuravy) [inline src/LAMMPS.jl:268]:
```suggestion
Create atoms for a LAMMPS instance. 
`x` contains the atom positions and should be a 3 by `n`,  `Matrix{Float64}`, where `n` is the number of atoms. 
id contains the id of each atom and should all be `Vector{Int32}` with length `n`.
types contains the atomic type (LAMMPS number) of each atom and should all be a `Vector{Int32}` with length `n`.
`v ` contains the associated velocities and should be a 3 by `n` Matrix{Float64}.
`image` contains the image flags for each atom and should be `Vector{IMAGEINT}` with length `n`.
`bexpand` is a `Bool` that defines whether or not the box should be expanded to fit the input atoms (default not).
```

Comment (jmeziere) [reply src/LAMMPS.jl]: I modified what you have here slightly. Let me know if I should change it back.

Comment (Joroks) [reply src/LAMMPS.jl]: There is an extra space after `v`

Comment (jmeziere) [reply src/LAMMPS.jl]: Ok, that last push should fix it.

File: test/runtests.jl
Comment (vchuravy) [inline test/runtests.jl:337]: These tests made me wonder if we could also provide a `view(image, 1:99)`

Comment (jmeziere) [reply test/runtests.jl:346]: Would we then have to allow for `SubArray` types to be passed in to the `create_atoms` function?

Comment (vchuravy) [reply test/runtests.jl:346]: Yeah I generally avoid being overly strict in the types.

Types in Julia are for dispatch and many types may fit into functions.

Comment (vchuravy) [reply test/runtests.jl:346]: But let's not block this PR on this decision, how to handle views etc coherently is a bigger question

Comment (jmeziere) [reply test/runtests.jl:346]: Got it. Is there anything you need me to do? It says that everything thing is good to merge, I believe.

**Pull request merged**
